### ISSUE TITLE
Machine setupSignals does not scale

### ISSUE THREAD

**Issue by cpheps** [firecracker-microvm/firecracker-go-sdk#198]
I just finished running some pprof analysis on our code and with the latest SDK that now has [machine.setupSignals](https://github.com/firecracker-microvm/firecracker-go-sdk/blob/master/machine.go#L939) I saw that the select statement eats up the majority of the total cpu time. Especially when the number of vms increases. 

I've included a screen shot of the firecracker portion of our cpu pprof
<img width="1082" alt="Screen Shot 2020-02-18 at 1 12 50 PM" src="https://user-images.githubusercontent.com/11877763/74764885-93473480-5250-11ea-8a5a-ca6880744ab8.png">

At peak we were running 55 vms on the same host and it was causing 1000% cpu usage for our program managing the vms via the SDK. If I set the `m.Cfg.ForwardSignals` slice to an empty slice performance increased dramatically, same number of vms but only 22% cpu usage for our program.

I did verify we were not leaking goroutines and that the setupSignals goroutine was being exited when the vm died.

**Comment by kzys** [comment]
Thank you for the profiling! We should have that somewhere in my pipeline.

The intention of the change (#166) is disabling signal handling in the SDK. So setting the slice empty would be the way to go.

However, I do want to make our default option performant. Can we give up the signal handling in the SDK? Signals are shared by a process. It doesn't fit nicely in the SDK.

Cc: @michaeldwan

**Comment by kzys** [comment]
According to @nmeyerhans, the original intention of the signal handling is having some sane default to avoid orphaning firecracker processes. So we should not remove the signal handling unless we have better alternatives.

**Comment by kzys** [comment]
@cpheps I've been trying to reproduce the issue. While I could get a similar pprof graph, ForwardSignals didn't affect the overall performance much.

The test code is currently in my fork: https://github.com/kzys/firecracker-go-sdk/commit/58ec39377b0c44378478df3d8c4148dbdcd502a7

On my z1d.metal EC2 instance, both TestRunMultipleVMWithSignalForwarding and TestRunMultipleVMWithoutSignalForwarding resulted similar results, in terms of `perf` numbers.

```
% uptime
 23:55:38 up 22 days,  5:27,  1 user,  load average: 0.00, 0.07, 5.71
% sudo perf stat -a -- sudo DISABLE_ROOT_TESTS= /usr/local/go/bin/go test -run=TestRunMultipleVMWithSignalForwarding
PASS
ok      github.com/firecracker-microvm/firecracker-go-sdk       30.092s

 Performance counter stats for 'system wide':

    1502907.983804      cpu-clock (msec)          #   47.997 CPUs utilized          
           155,507      context-switches          #    0.103 K/sec                  
             5,645      cpu-migrations            #    0.004 K/sec                  
           506,009      page-faults               #    0.337 K/sec                  
    69,748,593,803      cycles                    #    0.046 GHz                    
    31,811,500,826      instructions              #    0.46  insn per cycle         
     6,631,695,688      branches                  #    4.413 M/sec                  
       154,848,478      branch-misses             #    2.33% of all branches        

      31.312733676 seconds time elapsed

% sudo perf stat -a -- sudo DISABLE_ROOT_TESTS= /usr/local/go/bin/go test -run=TestRunMultipleVMWithoutSignalForwarding
PASS
ok      github.com/firecracker-microvm/firecracker-go-sdk       30.084s

 Performance counter stats for 'system wide':

    1488668.960217      cpu-clock (msec)          #   47.995 CPUs utilized          
           152,219      context-switches          #    0.102 K/sec                  
             5,735      cpu-migrations            #    0.004 K/sec                  
           482,550      page-faults               #    0.324 K/sec                  
    63,349,661,892      cycles                    #    0.043 GHz                    
    27,676,147,166      instructions              #    0.44  insn per cycle         
     5,806,599,701      branches                  #    3.901 M/sec                  
       137,671,057      branch-misses             #    2.37% of all branches        

      31.017347468 seconds time elapsed

%
```

**Comment by cpheps** [comment]
@kzys Took a look at your test and look like what I would expect. I did see one difference in that we set `ForwardSignals = []os.Signal{}` before we call `NewMachine`. Looking at the code this should be no difference. Do that though did improve performance for us and removed that issue.

**Comment by kzys** [comment]
Yeah. Setting that *before* calling NewMachine seems no-op because of https://github.com/firecracker-microvm/firecracker-go-sdk/blob/f8acfc43bfbe27b5c80264662f992573fdc0a20b/machine.go#L348.

How did you measure the performance? [CPU Utilization is something misleading](http://www.brendangregg.com/blog/2017-05-09/cpu-utilization-is-wrong.html). Did it actually increase the performance which is measured by wall-clock time?

**Comment by cpheps** [comment]
Right so my measurement was just watching `top` for our process. When the go routines in `setupSignals` were not created the cpu measurement in top went from 1000% to ~20%

**Comment by cpheps** [comment]
@kzys Did a quick read through I might know why you're not seeing improvements in your test. The Config is not a pointer that is passed to `NewMachine`

https://github.com/firecracker-microvm/firecracker-go-sdk/blob/f8acfc43bfbe27b5c80264662f992573fdc0a20b/machine.go#L303

So when it's set to the OS signals and set on the machine it's a modified copy of the one you still have a handle on in your tests

https://github.com/firecracker-microvm/firecracker-go-sdk/blob/f8acfc43bfbe27b5c80264662f992573fdc0a20b/machine.go#L348-L358

In your tests when you set `machine.Cfg.ForwardSignals = nil` you're only changing your local copy not the one the machine is using.

**Comment by kzys** [comment]
@cpheps Thanks. You are right! Let me check that again tomorrow.

**Comment by kzys** [comment]
@cpheps I can reproduce the issue and see the difference in wall-clock time. Yes. ForwardSignals does make the host busier.

```
% sudo perf stat -a -- sudo DISABLE_ROOT_TESTS= ~/src/go1.11.13/bin/go test -run=TestRunMultipleVMWithSignalForwarding                                       
starting 300 VMs
PASS
ok      github.com/firecracker-microvm/firecracker-go-sdk       4.326s

 Performance counter stats for 'system wide':

        532,034.11 msec cpu-clock                 #   95.889 CPUs utilized 
           234,664      context-switches          #  441.070 M/sec 
            14,358      cpu-migrations            #   26.987 M/sec 
         1,079,383      page-faults               # 2028.786 M/sec 
 1,155,639,302,716      cycles                    # 2172115.509 GHz 
 1,352,360,504,475      instructions              #    1.17  insn per cycle 
   317,303,636,622      branches                  # 596397291.568 M/sec 
     1,421,472,029      branch-misses             #    0.45% of all branches 

       5.548415648 seconds time elapsed

% sudo perf stat -a -- sudo DISABLE_ROOT_TESTS= ~/src/go1.11.13/bin/go test -run=TestRunMultipleVMWithoutSignalForwarding
starting 300 VMs
PASS
ok      github.com/firecracker-microvm/firecracker-go-sdk       0.939s

 Performance counter stats for 'system wide':

        208,201.10 msec cpu-clock                 #   95.664 CPUs utilized 
           147,282      context-switches          #  707.403 M/sec 
            12,860      cpu-migrations            #   61.767 M/sec 
         1,068,457      page-faults               # 5131.853 M/sec 
   110,600,490,679      cycles                    # 531219.786 GHz 
    53,971,928,336      instructions              #    0.49  insn per cycle 
    11,255,269,198      branches                  # 54059630.828 M/sec 
       236,858,202      branch-misses             #    2.10% of all branches 

       2.176368102 seconds time elapsed
```

**Comment by kzys** [comment]
@cpheps The SDK is not removing signal handling channels correctly. That means, we may be leaking goroutines. The fix is #204!

**Comment by cpheps** [comment]
Awesome! Thanks @kzys for chasing down a fix.
